src/cartesi-machine.lua

```diff
@@ -2185,6 +2199,12 @@ if max_uarch_cycle > 0 then
     end
 end
 if gdb_stub then gdb_stub:close() end
+if log_step_mcycle_count then
+    stderr(string.format("Logging step of %d cycles to %s\n", log_step_mcycle_count, log_step_filename))
+    print_root_hash(machine, stderr_unsilenceable)
```

Comment:
We need to print the hash before the call to log_step() because it occurs after the machine may have advanced a certain number of cycles. Similarly, printing the hash after is necessary because the machine can advance micro cycles following the logging of the step.

The --initial-hash and --final-hash flags only print the hashes of the very first and very last cycles.

It is reasonable to assume that users saving a step log would also want the corresponding hashes to validate the log elsewhere, hence the prints.

For example:
```
cartesi-machine.lua --initial-hash --max-mcycle=10 --log-step=1,/tmp/loxa1.bin --max-uarch-cycle=1 --final-hash
0: da95ad8352d6ff9183c8b3f7520c6a68097f42ef42ff3ddb52eb2238d325aad3
Logging step of 1 cycles to /tmp/loxa1.bin
10: ef886a04bc473480929ec2d14608286bf2705ed9a8767edcd30299131afc9f6f
11: b4182e651d2fa9036f255fc26c946d8432bd3b4ab50790415897e6d411b30a73
11: b4182e651d2fa9036f255fc26c946d8432bd3b4ab50790415897e6d411b30a73
```

Reply:
Hmmm. I we should throw if the uarch is not in a reset state when we do log-step. Mixing these two is basically a bug, right? Or some very adventurous person doing something way over their head. Can we do that?

Reply:
`log_step_uarch()` is always the last advance operation, when `--log-step-uarch`  is specified.

Reply:
I mean this combination `--log-step=1,/tmp/loxa1.bin --max-uarch-cycle=1` should not be acceptable.

Reply:
Got it.
The micro steps  need to occur after all big steps.
```
cartesi-machine.lua --initial-hash --max-mcycle=10 --log-step=1,/tmp/loxa1.bin --max-uarch-cycle=8000 --log-step-uarch --final-hash
0: da95ad8352d6ff9183c8b3f7520c6a68097f42ef42ff3ddb52eb2238d325aad3
Logging step of 1 cycles to /tmp/loxa1.bin
10: e9d613431892a7c6d4fafb64e828bf772c9583081de6fd8141943792909dd94d
11: 763ccef8428a0f68356ff421cdfcd4da2832160dd41fdac7d245c016b323424b
Cycles: 12
uCycles: 262
Gathering micro step log: please wait
begin step
  1: read uarch.cycle@0x400008(4194312): 0x106(262)
  2: read uarch.halt_flag@0x400000(4194304): 0x1(1)
end step
12: c75e0fd7e098ba66baf20d0baa2ec7a2452003431334b587d7ccc5079c613e49
``